### PR TITLE
8309203: C2: remove copy-by-value of GrowableArray for InterfaceSet

### PR DESCRIPTION
**Pull request by rwestrel**
The `InterfaceSet` component of types is immutable (once it is fully
constructed). In that way, it's similar to other type components. In
this change I propose making `InterfaceSet` a subtype of `Type` (and
renaming it `TypeInterfaces`). It's then feasible to `hashcons`
`TypeInterfaces` instances the way it's done for other `Type`
instance.  A single copy of a `TypeInterfaces` for a particular set of
interfaces is live during a compilation (and for the entire length of
the compilation) and pointer equality can be used to check two sets of
interfaces for equality. I think that change makes interface support
fit better with the type system implementation and has the added
benefits that there's no copy by value of the GrowableArray that holds
the set of interfaces anymore.

A couple of required virtual methods for `Type` are not implemented
for `TypeInterfaces`: `xmeet` and `singleton` (their body now calls
`ShouldNotReachHere`). They would need a way to tell whether
`TypeInterfaces` instances are above the center line or not. I thought
about (and tried) having a root class `Type` that would contain the
bare minimum for type that can be `hashcons`'ed so this method
wouldn't need to be implemented by `TypeInterfaces`. That proved to be
quite complicated and I gave up.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309203](https://bugs.openjdk.org/browse/JDK-8309203): C2: remove copy-by-value of GrowableArray for InterfaceSet (**Sub-task** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16309/head:pull/16309` \
`$ git checkout pull/16309`

Update a local copy of the PR: \
`$ git checkout pull/16309` \
`$ git pull https://git.openjdk.org/jdk.git pull/16309/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16309`

View PR using the GUI difftool: \
`$ git pr show -t 16309`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16309.diff">https://git.openjdk.org/jdk/pull/16309.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16309#issuecomment-1775118311)

### CODE REVIEW

**Comment by rwestrel** [comment]
> I'm seeing a build failure:

Should be fixed now.

File: src/hotspot/share/opto/type.cpp [hunk 1]
```diff
@@ -63,6 +63,8 @@ const Type::TypeInfo Type::_type_info[Type::lastype] = {
   { Bad,             T_ILLEGAL,    "tuple:",        false, Node::NotAMachineReg, relocInfo::none          },  // Tuple
   { Bad,             T_ARRAY,      "array:",        false, Node::NotAMachineReg, relocInfo::none          },  // Array
 
+  { Bad,             T_ARRAY,      "interfaces:",  false, Node::NotAMachineReg, relocInfo::none          },  // Interfaces
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.cpp:66]: Indentation is wrong.

File: src/hotspot/share/opto/type.cpp [hunk 2]
```diff
@@ -3253,28 +3255,33 @@ void TypeRawPtr::dump2( Dict &d, uint depth, outputStream *st ) const {
...
           _hash(0), _exact_klass(nullptr) {
   for (int i = 0; i < interfaces->length(); i++) {
     add(interfaces->at(i));
   }
   initialize();
 }
 
-void TypePtr::InterfaceSet::initialize() {
+const TypeInterfaces* TypeInterfaces::make(GrowableArray<ciInstanceKlass*>* interfaces) {
+  TypeInterfaces* result = interfaces == nullptr ? new TypeInterfaces() : new TypeInterfaces(interfaces);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.cpp:3274]:
```suggestion
  TypeInterfaces* result = (interfaces == nullptr) ? new TypeInterfaces() : new TypeInterfaces(interfaces);
```

File: src/hotspot/share/opto/type.cpp [hunk 3]
```diff
@@ -3325,12 +3328,16 @@ bool TypePtr::InterfaceSet::eq(ciInstanceKlass* k) const {
...
 
 
-uint TypePtr::InterfaceSet::hash() const {
+uint TypeInterfaces::hash() const {
   assert(_initialized, "must be");
   return _hash;
 }
 
-void TypePtr::InterfaceSet::compute_hash() {
+const Type *TypeInterfaces::xdual() const {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.cpp:3336]:
```suggestion
const Type* TypeInterfaces::xdual() const {
```

File: src/hotspot/share/opto/type.cpp [hunk 4]
```diff
@@ -3374,93 +3381,93 @@ void TypePtr::InterfaceSet::verify() const {
...
   int j = 0;
-  while (i < _list.length() || j < other._list.length()) {
+  while (i < _list.length() || j < other->_list.length()) {
     while (i < _list.length() &&
-           (j >= other._list.length() ||
-            compare(_list.at(i), other._list.at(j)) < 0)) {
-      result.raw_add(_list.at(i));
+           (j >= other->_list.length() ||
+            compare(_list.at(i), other->_list.at(j)) < 0)) {
+      result_list.push((ciInstanceKlass*)_list.at(i));
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.cpp:3392]: I think you should use `as_instance_klass()` here and in other code.

File: src/hotspot/share/opto/type.hpp [hunk 5]
```diff
@@ -94,6 +94,8 @@ class Type {
     Tuple,                      // Method signature or object layout
     Array,                      // Array types
 
+    Interfaces,
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.hpp:97]: Please add a comment

File: src/hotspot/share/opto/type.hpp [hunk 6]
```diff
@@ -872,6 +874,47 @@ class TypeVectMask : public TypeVect {
   static const TypeVectMask* make(const Type* elem, uint length);
 };
 
+class TypeInterfaces : public Type {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.hpp:877]: Please add a comment

**Review by TobiHartmann — APPROVED** [review]
Very nice cleanup. I just added a few style comments.

File: src/hotspot/share/opto/type.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/type.cpp]: fixed in new commit.

File: src/hotspot/share/opto/type.cpp (hunk 4 above)
Comment (rwestrel) [reply src/hotspot/share/opto/type.cpp]: I moved both `_list` and `result_list` to `ciInstanceKlass` so there's no cast anymore.

**Comment by rwestrel** [comment]
> Very nice cleanup. I just added a few style comments.

Thanks for reviewing this. All your comments should be addressed now.

**Review by TobiHartmann — APPROVED** [review]
Thanks, looks good!

**Comment by rwestrel** [comment]
Anyone else for a review of this change?

**Review by vnkozlov — APPROVED** [review]
Good. cleanup.

**Comment by rwestrel** [comment]
@TobiHartmann @vnkozlov thanks for the reviews.

**Pull request closed without merging**